entgql/plugin/relay.go

```diff
@@ -0,0 +1,121 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+
+	"entgo.io/ent/entc/gen"
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+var (
+	RelayCursor   = "Cursor"
+	RelayNode     = "Node"
+	RelayPageInfo = "PageInfo"
+)
+
+func (e *EntGQL) relayBuiltins() {
+	e.insertDefinitions([]*ast.Definition{
+		{
+			Name: RelayCursor,
+			Kind: ast.Scalar,
+			Description: `Define a Relay Cursor type:
+https://relay.dev/graphql/connections.htm#sec-Cursor`,
+		},
+		{
+			Kind: ast.Interface,
+			Name: RelayNode,
+			Description: `An object with an ID.
+Follows the [Relay Global Object Identification Specification](https://relay.dev/graphql/objectidentification.htm)`,
+			Fields: []*ast.FieldDefinition{
+				{
+					Name:        "id",
+					Type:        ast.NonNullNamedType("ID", nil),
+					Description: "The id of the object.",
+				},
+			},
+		},
+		{
+			Name: RelayPageInfo,
+			Kind: ast.Object,
+			Description: `Information about pagination in a connection.
+https://relay.dev/graphql/connections.htm#sec-undefined.PageInfo`,
+			Fields: []*ast.FieldDefinition{
+				{
+					Name:        "hasNextPage",
+					Type:        ast.NonNullNamedType("Boolean", nil),
+					Description: "When paginating forwards, are there more items?",
+				},
+				{
+					Name:        "hasPreviousPage",
+					Type:        ast.NonNullNamedType("Boolean", nil),
+					Description: "When paginating backwards, are there more items?",
+				},
+				{
+					Name:        "startCursor",
+					Type:        ast.NamedType("Cursor", nil),
+					Description: "When paginating backwards, the cursor to continue.",
+				},
+				{
+					Name:        "endCursor",
+					Type:        ast.NamedType("Cursor", nil),
+					Description: "When paginating forwards, the cursor to continue.",
+				},
+			},
+		},
+	})
+}
+
+func (e *EntGQL) relayConnection(t *gen.Type) {
+	e.insertDefinition(&ast.Definition{
+		Name:        fmt.Sprintf("%sEdge", t.Name),
+		Kind:        ast.Object,
+		Description: "An edge in a connection.",
+		Fields: []*ast.FieldDefinition{
+			{
+				Name:        "node",
+				Type:        ast.NamedType(t.Name, nil),
+				Description: "The item at the end of the edge",
+			},
+			{
+				Name:        "cursor",
+				Type:        ast.NonNullNamedType("Cursor", nil),
+				Description: "A cursor for use in pagination",
+			},
+		},
+	})
+	e.insertDefinition(&ast.Definition{
+		Name:        fmt.Sprintf("%sConnection", t.Name),
+		Kind:        ast.Object,
+		Description: "A connection to a list of items.",
+		Fields: []*ast.FieldDefinition{
+			{
+				Name:        "edges",
+				Type:        ast.ListType(ast.NamedType(fmt.Sprintf("%sEdge", t.Name), nil), nil),
+				Description: "A list of edges.",
+			},
+			{
+				Name:        "pageInfo",
+				Type:        ast.NonNullNamedType("PageInfo", nil),
+				Description: "Information to aid in pagination.",
+			},
+			{
+				Name: "totalCount",
```

Comment:
nit: this should be configurable

Reply:
Is there any reason that we should not expose this field? It's not defined in the relay spec but it's exists in `graphql-relay-js`

Reply:
just to give the user the option to hide it if they wanted too